File: 32_03_PrintTreesInZigzag/PrintTreesInZigzag.cpp
#include <iostream>
#include <vector>
#include <string>
#include <deque>
#include <set>
#include <map>
#include <queue>
#include <stack>
#include <algorithm>
#include <cstring>
using namespace std;

struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution
{
  public:
    vector<vector<int>> Print(TreeNode *pRoot)
    {
        vector<vector<int>> res;
        if (pRoot == nullptr)
            return res;
        stack<TreeNode *> leftStack, rightStack;
        leftStack.push(pRoot);
        while (!leftStack.empty() || !rightStack.empty())
        {
            if (!leftStack.empty())
            {
                vector<int> temp;
                while (!leftStack.empty())
                {
                    TreeNode *data = leftStack.top();
                    leftStack.pop();
                    temp.push_back(data->val);
                    if (data->left)
                        rightStack.push(data->left);
                    if (data->right)
                        rightStack.push(data->right);
                }
                res.push_back(temp);
            }
            if (!rightStack.empty())
            {
                vector<int> temp;
                while (!rightStack.empty())
                {
                    TreeNode *data = rightStack.top();
                    rightStack.pop();
                    temp.push_back(data->val);
                    if (data->right)
                        leftStack.push(data->right);
                    if (data->left)
                        leftStack.push(data->left);
                }
                res.push_back(temp);
            }
        }
        return res;
    }
};

File: 32_03_PrintTreesInZigzag/PrintTreesInZigzag.py
from collections import deque


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


def reConstructBinaryTree(pre, lin):
    if not pre and not lin:
        return None
    root = TreeNode(pre[0])
    if set(pre) != set(lin):
        return None
    i = lin.index(pre[0])
    root.left = reConstructBinaryTree(pre[1:i + 1], lin[:i])
    root.right = reConstructBinaryTree(pre[i + 1:], lin[i + 1:])
    return root


class Solution:
    def Print(self, pRoot):
        # write code here
        if not pRoot:
            return []

        res, tmp = [], []
        last = pRoot
        q = deque([pRoot])
        left_to_right = True
        while q:
            t = q.popleft()
            tmp.append(t.val)
            if t.left:
                q.append(t.left)
            if t.right:
                q.append(t.right)
            if t == last:
                res.append(tmp if left_to_right else tmp[::-1])
                left_to_right = not left_to_right
                tmp = []
                if q:
                    last = q[-1]
        return res


t = reConstructBinaryTree([1, 2, 4, 8, 9, 5, 10, 11, 3, 6, 12, 13, 7, 14, 15],
                          [8, 4, 9, 2, 10, 5, 11, 1, 12, 6, 13, 3, 14, 7, 15])
s = Solution()
print(s.Print(t))